### PR TITLE
cephadm: log to file

### PR DESCRIPTION
**Pull request by jmolmo**
Log configuration improved adding a log file with debug level.

Now, in the new log file we can see something like:

```
[root@cephLab2-node-00 bin]# cat /var/log/ceph/cephadm.log 
--------------------------------------------------------------------------------
cephadm ['version']
Running command: podman images --filter label=ceph=True --format {{.Repository}} {{.Tag}}
podman:stdout docker.io/ceph/daemon-base latest-master-devel
podman:stdout <none> 
Using recent ceph image docker.io/ceph/daemon-base:latest-master-devel
['podman', 'run', '--rm', '--net=host', '--ipc=host', '-e', 'CONTAINER_IMAGE=docker.io/ceph/daemon-base:latest-master-devel', '-e', 'NODE_NAME=cephLab2-node-00', '--entrypoint', 'ceph', 'docker.io/ceph/daemon-base:latest-master-devel', '--version']
Running command: podman run --rm --net=host --ipc=host -e CONTAINER_IMAGE=docker.io/ceph/daemon-base:latest-master-devel -e NODE_NAME=cephLab2-node-00 --entrypoint ceph docker.io/ceph/daemon-base:latest-master-devel --version
ceph:stdout ceph version 16.0.0-4623-g8db888ea42 (8db888ea4210641b838d21631c25fa56752fe970) pacific (dev)
--------------------------------------------------------------------------------
cephadm ['ls']
Running command: systemctl is-enabled ceph-90bdc44c-e5f6-11ea-bd53-525400893115@grafana.cephLab2-node-00
systemctl:stdout enabled
Running command: systemctl is-active ceph-90bdc44c-e5f6-11ea-bd53-525400893115@grafana.cephLab2-node-00
systemctl:stdout active
Running command: podman --version
...

Running command: podman inspect --format {{.Id}},{{.Config.Image}},{{.Image}},{{.Created}},{{index .Config.Labels "io.ceph.version"}} ceph-90bdc44c-e5f6-11ea-bd53-525400893115-mon.cephLab2-node-00
podman:stdout 78160f922fe19e015780fca1a692b4136cb4a13a57507bf63d5f4c29db4cde34,docker.io/ceph/daemon-base:latest-master-devel,8b9b100f95caef2b3d5cdf5cddd1f1a92e4e731e3ee7e0d8dc91ad54d3d75d50,2020-08-24 10:43:03.351240858 +0000 UTC,
--------------------------------------------------------------------------------
cephadm ['rm-cluster', '--fsid', '90bdc44c-e5f6-11ea-bd53-525400893115', '--force']
Acquiring lock 139638194999104 on /run/cephadm/90bdc44c-e5f6-11ea-bd53-525400893115.lock
Lock 139638194999104 acquired on /run/cephadm/90bdc44c-e5f6-11ea-bd53-525400893115.lock
Running command: systemctl stop ceph-90bdc44c-e5f6-11ea-bd53-525400893115@grafana.cephLab2-node-00
Running command: systemctl reset-failed ceph-90bdc44c-e5f6-11ea-bd53-525400893115@grafana.cephLab2-node-00
Running command: systemctl disable ceph-90bdc44c-e5f6-11ea-bd53-525400893115@grafana.cephLab2-node-00
...

Running command: rm -f /etc/logrotate.d/ceph-90bdc44c-e5f6-11ea-bd53-525400893115
Releasing lock 139638194999104 on /run/cephadm/90bdc44c-e5f6-11ea-bd53-525400893115.lock
Lock 139638194999104 released on /run/cephadm/90bdc44c-e5f6-11ea-bd53-525400893115.lock
--------------------------------------------------------------------------------
cephadm ['ls']
```

Signed-off-by: Juan Miguel Olmo Martínez <jolmomar@redhat.com>

### CODE REVIEW

**Review by sebastian-philipp — APPROVED** [review]

**Comment by sebastian-philipp** [comment]
I think this log file is already properly archived: https://github.com/ceph/ceph/blob/master/qa/tasks/cephadm.py#L157 .

File: src/cephadm/cephadm [hunk 1]
```diff
@@ -44,6 +44,7 @@ import fcntl
 import ipaddress
 import json
 import logging
+from logging.config import dictConfig
```
Comment (mgfritch) [inline src/cephadm/cephadm:47]:
```suggestion
import logging
```

import twice?

File: src/cephadm/cephadm [hunk 2]
```diff
@@ -92,6 +93,32 @@ cached_stdin = None
...
+        }
+    },
+    'loggers': {
+        '': {
+            'level': 'DEBUG',
+            'handlers': ['console', 'log_file'],
+        }
+    }
+}
+dictConfig(logging_config)
```
Comment (mgfritch) [inline src/cephadm/cephadm:120]:
```suggestion
logging.dictConfig(logging_config)
```

File: src/cephadm/cephadm (hunk 1 above)
Comment (mgfritch) [reply src/cephadm/cephadm:47]: nm, I misread this, it's a separate module ..

**Comment by jmolmo** [comment]
Unit test was failing because the creation of the log file.

I have mock the file open and removed "py27" from the tests. (AFAIK cephadm is a python 3 tool for work in systems with Python 3.)

**Comment by sebastian-philipp** [comment]
cephadm is not yet a py3 tool. There are places in Teuthology where the compatibility to py2 is verified. @pcuzner is looking into them right now.

**Comment by sebastian-philipp** [comment]
```
2020-08-25T12:38:05.841 INFO:teuthology.orchestra.run.smithi097.stderr:Traceback (most recent call last):
2020-08-25T12:38:05.841 INFO:teuthology.orchestra.run.smithi097.stderr:  File "/usr/lib64/python3.6/logging/config.py", line 565, in configure
2020-08-25T12:38:05.841 INFO:teuthology.orchestra.run.smithi097.stderr:    handler = self.configure_handler(handlers[name])
2020-08-25T12:38:05.842 INFO:teuthology.orchestra.run.smithi097.stderr:  File "/usr/lib64/python3.6/logging/config.py", line 738, in configure_handler
2020-08-25T12:38:05.842 INFO:teuthology.orchestra.run.smithi097.stderr:    result = factory(**kwargs)
2020-08-25T12:38:05.842 INFO:teuthology.orchestra.run.smithi097.stderr:  File "/usr/lib64/python3.6/logging/handlers.py", line 150, in __init__
2020-08-25T12:38:05.842 INFO:teuthology.orchestra.run.smithi097.stderr:    BaseRotatingHandler.__init__(self, filename, mode, encoding, delay)
2020-08-25T12:38:05.842 INFO:teuthology.orchestra.run.smithi097.stderr:  File "/usr/lib64/python3.6/logging/handlers.py", line 57, in __init__
2020-08-25T12:38:05.843 INFO:teuthology.orchestra.run.smithi097.stderr:    logging.FileHandler.__init__(self, filename, mode, encoding, delay)
2020-08-25T12:38:05.843 INFO:teuthology.orchestra.run.smithi097.stderr:  File "/usr/lib64/python3.6/logging/__init__.py", line 1032, in __init__
2020-08-25T12:38:05.843 INFO:teuthology.orchestra.run.smithi097.stderr:    StreamHandler.__init__(self, self._open())
2020-08-25T12:38:05.843 INFO:teuthology.orchestra.run.smithi097.stderr:  File "/usr/lib64/python3.6/logging/__init__.py", line 1061, in _open
2020-08-25T12:38:05.844 INFO:teuthology.orchestra.run.smithi097.stderr:    return open(self.baseFilename, self.mode, encoding=self.encoding)
```

https://pulpito.ceph.com/swagner-2020-08-25_11:22:22-rados:cephadm-wip-swagner2-testing-2020-08-25-1024-distro-basic-smithi/5375706/

**Comment by smithfarm** [comment]
Would it make sense to change all `print` statements to `logger.info`? It seems like the presence of `print` statements means that there is at least a risk that some information will appear only on the console and not be logged to a file.

**Comment by sebastian-philipp** [comment]
using `print` makes sure that the message will never be altered. if we print some (json) output, we don't want any other junk in the output.

**Comment by smithfarm** [comment]
> using `print` makes sure that the message will never be altered. if we print some (json) output, we don't want any other junk in the output.

That's a very good reason to use print. Still, it seems like it might be useful, when we print something, to also log what we are printing.

**Comment by sebastian-philipp** [comment]
green: https://pulpito.ceph.com/swagner-2020-08-27_07:38:41-rados:cephadm-wip-swagner2-testing-2020-08-26-1329-distro-basic-smithi/

**Comment by sebastian-philipp** [comment]
http://qa-proxy.ceph.com/teuthology/swagner-2020-08-27_07:38:41-rados:cephadm-wip-swagner2-testing-2020-08-26-1329-distro-basic-smithi/5379643/remote/ubuntu%40smithi051.front.sepia.ceph.com/log/cephadm.log.gz

looks like

```
cephadm ['--image', 'quay.ceph.io/ceph-ci/ceph:30b2598f1697771f1d65eb304cb07af80fd21fc3', 'deploy', '--fsid', 'd82874fe-e83e-11ea-a073-001a4aab830c', '--name', 'mon.a', '--config-json', '-', '--reconfig']
Acquiring lock 140572358387192 on /run/cephadm/d82874fe-e83e-11ea-a073-001a4aab830c.lock
Lock 140572358387192 not acquired on /run/cephadm/d82874fe-e83e-11ea-a073-001a4aab830c.lock, waiting 0.05 seconds ...
...snip...
/bin/podman:profile rt=1.533416748046875, stop=False, exit=None, reads=[5]
/bin/podman:profile rt=1.5337121486663818, stop=False, exit=None, reads=[5]
/bin/podman:stdout ...snip...
/bin/podman:profile rt=2.0273022651672363, stop=False, exit=0, reads=[5, 7]
/bin/podman:profile rt=2.0274717807769775, stop=True, exit=0, reads=[5, 7]
Releasing lock 140399422237552 on /run/cephadm/d82874fe-e83e-11ea-a073-001a4aab830c.lock
Lock 140399422237552 released on /run/cephadm/d82874fe-e83e-11ea-a073-001a4aab830c.lock
Acquiring lock 140572358387192 on /run/cephadm/d82874fe-e83e-11ea-a073-001a4aab830c.lock
Lock 140572358387192 acquired on /run/cephadm/d82874fe-e83e-11ea-a073-001a4aab830c.lock
Running command: systemctl is-enabled ceph-d82874fe-e83e-11ea-a073-001a4aab830c@mon.a
systemctl:stdout enabled
Running command: systemctl is-active ceph-d82874fe-e83e-11ea-a073-001a4aab830c@mon.a
systemctl:stdout active
Reconfig daemon mon.a ...
['/bin/podman', 'run', '--rm', '--net=host', '--ipc=host', '-e', 'CO...snip...
Running command: /bin/podman run --rm --net=host --ipc=host -e CONTAIN...snip...
stat:stdout 167 167
Running command: install -d -m0770 -o 167 -g 167 /var/run/ceph/d82874fe-e83e-11ea-a073-001a4aab830c
Running command: systemctl is-enabled firewalld.service
systemctl:stdout disabled
Running command: systemctl is-active firewalld.service
systemctl:stdout inactive
firewalld.service is not enabled
Not possible to enable service <mon>. firewalld.service is not available
Not possible to open ports <[]>. firewalld.service is not available
Releasing lock 140572358387192 on /run/cephadm/d82874fe-e83e-11ea-a073-001a4aab830c.lock
Lock 140572358387192 released on /run/cephadm/d82874fe-e83e-11ea-a073-001a4aab830c.lock
```

Do you want to add some timestamp + level prefix to the file?

**Comment by jmolmo** [comment]
> Do you want to add some timestamp + level prefix to the file?

Sure! Thanks  for the suggestion really it was needed :-)

Now it looks like:

```
[root@cephLab2-node-00 ~]# cat /var/log/ceph/cephadm.log 
2020-08-31 10:31:48,677 DEBUG --------------------------------------------------------------------------------
cephadm ['version']
2020-08-31 10:31:48,685 DEBUG Running command: /usr/bin/podman images --filter label=ceph=True --format {{.Repository}} {{.Tag}}
2020-08-31 10:31:48,943 WARNING This is a development version of cephadm.
2020-08-31 10:31:48,944 WARNING For information regarding the latest stable release:
2020-08-31 10:31:48,944 WARNING     https://docs.ceph.com/docs/octopus/cephadm/install
2020-08-31 10:31:48,944 DEBUG ['/usr/bin/podman', 'run', '--rm', '--net=host', '--ipc=host', '-e', 'CONTAINER_IMAGE=docker.io/ceph/daemon-base:latest-master-devel', '-e', 'NODE_NAME=cephLab2-node-00', '--entrypoint', 'ceph', 'docker.io/ceph/daemon-base:latest-master-devel', '--version']
2020-08-31 10:31:48,944 DEBUG Running command: /usr/bin/podman run --rm --net=host --ipc=host -e CONTAINER_IMAGE=docker.io/ceph/daemon-base:latest-master-devel -e NODE_NAME=cephLab2-node-00 --entrypoint ceph docker.io/ceph/daemon-base:latest-master-devel --version
2020-08-31 10:31:49,025 DEBUG ceph:stderr Trying to pull docker.io/ceph/daemon-base:latest-master-devel...
2020-08-31 10:31:50,933 DEBUG ceph:stderr Getting image source signatures
2020-08-31 10:31:51,409 DEBUG ceph:stderr Copying blob sha256:3c72a8ed68140139e483fe7368ae4d9651422749e91483557cbd5ecf99a96110
2020-08-31 10:31:51,429 DEBUG ceph:stderr Copying blob sha256:d78aab09c054b206910d7a6d8409574c96e8355e7bd18c0414bc30857095c800
2020-08-31 10:32:02,315 DEBUG ceph:stderr Copying config sha256:3843661e27d60dbb33b2aae30b782fcd33f42533a4305b978257b9bac8b0aa2b
2020-08-31 10:32:02,320 DEBUG ceph:stderr Writing manifest to image destination
2020-08-31 10:32:02,321 DEBUG ceph:stderr Storing signatures
2020-08-31 10:32:18,656 DEBUG ceph:stdout ceph version 16.0.0-4878-g90e5434905 (90e54349056202e5442a8e57c5752da4ea85225b) pacific (dev)
2020-08-31 10:34:11,202 DEBUG --------------------------------------------------------------------------------
cephadm ['ls']
2020-08-31 10:34:21,653 DEBUG --------------------------------------------------------------------------------
cephadm ['..help']
2020-08-31 10:34:25,126 DEBUG --------------------------------------------------------------------------------
cephadm ['--help']
```

**Comment by sebastian-philipp** [comment]
```
==================================== ERRORS ====================================
____________________ ERROR collecting tests/test_cephadm.py ____________________
tests/test_cephadm.py:13: in <module>
    cd = SourceFileLoader('cephadm', 'cephadm').load_module()
<frozen importlib._bootstrap_external>:399: in _check_name_wrapper
    ???
<frozen importlib._bootstrap_external>:823: in load_module
    ???
<frozen importlib._bootstrap_external>:682: in load_module
    ???
<frozen importlib._bootstrap>:265: in _load_module_shim
    ???
<frozen importlib._bootstrap>:684: in _load
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
<frozen importlib._bootstrap_external>:678: in exec_module
    ???
<frozen importlib._bootstrap>:219: in _call_with_frames_removed
    ???
cephadm:128: in <module>
    os.makedirs(LOG_DIR)
/usr/lib/python3.6/os.py:220: in makedirs
    mkdir(name, mode)
E   PermissionError: [Errno 13] Permission denied: '/var/log/ceph'
```
https://jenkins.ceph.com/job/ceph-pull-requests/58813/consoleFull#-203499223e840cee4-f4a4-4183-81dd-42855615f2c1

**Comment by sebastian-philipp** [comment]
green: https://pulpito.ceph.com/swagner-2020-08-31_14:44:13-rados:cephadm-wip-swagner3-testing-2020-08-31-1454-distro-basic-smithi/

File: src/cephadm/cephadm [hunk 3]
```diff
@@ -93,6 +94,40 @@ cached_stdin = None
...
+    },
+    'loggers': {
+        '': {
+            'level': 'DEBUG',
+            'handlers': ['console', 'log_file'],
+        }
+    }
+}
+if not os.path.exists(LOG_DIR):
+    os.makedirs(LOG_DIR)
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm:128]: you're creating the log folder at import time. might make sense to put this into `__name__ == '__main__'`?

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]

**Comment by sebastian-philipp** [comment]
ping. I'd like to get this in!

**Comment by sebastian-philipp** [comment]
ping?

File: src/cephadm/cephadm (hunk 3 above)
Comment (jmolmo) [reply src/cephadm/cephadm]: Moved!

**Comment by sebastian-philipp** [comment]
Failed:

```
2020-09-07T12:41:35.334 INFO:teuthology.orchestra.run.smithi146.stderr:Traceback (most recent call last):
2020-09-07T12:41:35.334 INFO:teuthology.orchestra.run.smithi146.stderr:  File "/home/ubuntu/cephtest/cephadm", line 5823, in <module>
2020-09-07T12:41:35.334 INFO:teuthology.orchestra.run.smithi146.stderr:    r = args.func()
2020-09-07T12:41:35.334 INFO:teuthology.orchestra.run.smithi146.stderr:  File "/home/ubuntu/cephtest/cephadm", line 1147, in _infer_fsid
2020-09-07T12:41:35.335 INFO:teuthology.orchestra.run.smithi146.stderr:    return func()
2020-09-07T12:41:35.335 INFO:teuthology.orchestra.run.smithi146.stderr:  File "/home/ubuntu/cephtest/cephadm", line 1184, in _infer_config
2020-09-07T12:41:35.335 INFO:teuthology.orchestra.run.smithi146.stderr:    return func()
2020-09-07T12:41:35.335 INFO:teuthology.orchestra.run.smithi146.stderr:  File "/home/ubuntu/cephtest/cephadm", line 1230, in _infer_image
2020-09-07T12:41:35.335 INFO:teuthology.orchestra.run.smithi146.stderr:    return func()
2020-09-07T12:41:35.335 INFO:teuthology.orchestra.run.smithi146.stderr:  File "/home/ubuntu/cephtest/cephadm", line 3258, in command_shell
2020-09-07T12:41:35.336 INFO:teuthology.orchestra.run.smithi146.stderr:    make_log_dir(args.fsid)
2020-09-07T12:41:35.336 INFO:teuthology.orchestra.run.smithi146.stderr:  File "/home/ubuntu/cephtest/cephadm", line 1324, in make_log_dir
2020-09-07T12:41:35.336 INFO:teuthology.orchestra.run.smithi146.stderr:    uid, gid = extract_uid_gid()
2020-09-07T12:41:35.336 INFO:teuthology.orchestra.run.smithi146.stderr:  File "/home/ubuntu/cephtest/cephadm", line 1911, in extract_uid_gid
2020-09-07T12:41:35.336 INFO:teuthology.orchestra.run.smithi146.stderr:    raise RuntimeError('uid/gid not found')
2020-09-07T12:41:35.337 INFO:teuthology.orchestra.run.smithi146.stderr:RuntimeError: uid/gid not found
2020-09-07T12:41:35.350 DEBUG:teuthology.orchestra.run:got remote process result: 1
```

https://pulpito.ceph.com/swagner-2020-09-07_12:14:05-rados:cephadm-wip-swagner-testing-2020-09-07-1054-distro-basic-smithi/5415702/

**Comment by jmolmo** [comment]
@sebastian-philipp : 
I think that the error produced in theutology has nothing to do with the code introduced in this pull request. 
If you take a look to the log file provided by the Theutology test ( just a little bit above of the extract provided above) you can see:

```
2020-09-07T12:41:16.884 INFO:teuthology.orchestra.run.smithi146:> sudo /home/ubuntu/cephtest/cephadm --image quay.ceph.io/ceph-ci/ceph:562c164a6b9a7162e805cfa3d1945ca448867e55 shell -c /etc/ceph/ceph.conf -k /etc/ceph/ceph.client.admin.keyring --fsid b987fc56-f106-11ea-a080-001a4aab830c -- ceph orch host add smithi146
2020-09-07T12:41:35.330 INFO:teuthology.orchestra.run.smithi146.stderr:Non-zero exit code 125 from /bin/podman run --rm --net=host --ipc=host -e CONTAINER_IMAGE=quay.ceph.io/ceph-ci/ceph:562c164a6b9a7162e805cfa3d1945ca448867e55 -e NODE_NAME=smithi146 --entrypoint stat quay.ceph.io/ceph-ci/ceph:562c164a6b9a7162e805cfa3d1945ca448867e55 -c %u %g /var/lib/ceph
2020-09-07T12:41:35.331 INFO:teuthology.orchestra.run.smithi146.stderr:stat:stderr Trying to pull quay.ceph.io/ceph-ci/ceph:562c164a6b9a7162e805cfa3d1945ca448867e55...
2020-09-07T12:41:35.331 INFO:teuthology.orchestra.run.smithi146.stderr:stat:stderr   can't talk to a V1 docker registry
2020-09-07T12:41:35.332 INFO:teuthology.orchestra.run.smithi146.stderr:stat:stderr Error: unable to pull quay.ceph.io/ceph-ci/ceph:562c164a6b9a7162e805cfa3d1945ca448867e55: unable to pull image: Error initializing source docker://quay.ceph.io/ceph-ci/ceph:562c164a6b9a7162e805cfa3d1945ca448867e55: can't talk to a V1 docker registry
2020-09-07T12:41:35.334 INFO:teuthology.orchestra.run.smithi146.stderr:Traceback (most recent call last):
2020-09-07T12:41:35.334 INFO:teuthology.orchestra.run.smithi146.stderr:  File "/home/ubuntu/cephtest/cephadm", line 5823, in <module>
2020-09-07T12:41:35.334 INFO:teuthology.orchestra.run.smithi146.stderr:    r = args.func()
...
```
The error is produced because there is no way to obtain the uid and gid in the function "extract_uid_gid" , because to get this information is needed to run a container and it was impossible to download the image.

**Comment by sebastian-philipp** [comment]
green https://pulpito.ceph.com/swagner-2020-09-10_12:00:54-rados:cephadm-wip-swagner4-testing-2020-09-10-1104-distro-basic-smithi/

**Comment by sebastian-philipp** [comment]
I'm a bit worried that failed-to-pull errors will now be hidden behind an error that looks like something is wrong with the logging.

**Review by sebastian-philipp — APPROVED** [review]

**Comment by sebastian-philipp** [comment]
(whoops. Missed the reviewed-by lines.)

**Pull request merged**